File: solutions/dart/food-chain/1/lib/food_chain.dart
// Dedicated to Shree DR.MDD

typedef Insect = ({String genus, String? quote});

class FoodChain {
  static final fauna = const <Insect>[
    (genus: 'fly', quote: null),
    (genus: 'spider', quote: 'It wriggled and jiggled and tickled inside her.'),
    (genus: 'bird', quote: 'How absurd to swallow a bird!'),
    (genus: 'cat', quote: 'Imagine that, to swallow a cat!'),
    (genus: 'dog', quote: 'What a hog, to swallow a dog!'),
    (genus: 'goat', quote: 'Just opened her throat and swallowed a goat!'),
    (genus: 'cow', quote: "I don't know how she swallowed a cow!"),
    (genus: 'horse', quote: "She's dead, of course!"),
  ];

  String intro(String genus) => 'I know an old lady who swallowed a $genus.';

  String consumed(String eater, String meal) =>
      'She swallowed the $eater to catch the $meal.';

  static final finalLine =
      "I don't know why she swallowed the fly. Perhaps she'll die.";

  String describe(String genus) => switch (genus) {
        'spider' => 'spider that wriggled and jiggled and tickled inside her',
        _ => genus
      };

  bool terminator(String genus) => genus == 'horse';

  List<String> recite(int start, int end) {
    var lines = <String>[];
    for (var idx = start - 1; idx < end; idx++) {
      lines.add('');
      lines.addAll(render(idx));
    }
    return lines.sublist(1);
  }

  List<String> render(int idx) {
    var stanza = <String>[];
    stanza.add(intro(fauna[idx].genus));

    if (fauna[idx].quote case var line?) stanza.add(line);
    if (terminator(fauna[idx].genus)) return stanza;

    for (var k = idx - 1; k >= 0; k--) {
      stanza.add(consumed(fauna[k + 1].genus, describe(fauna[k].genus)));
    }

    stanza.add(finalLine);
    return stanza;
  }
}

File: solutions/ruby/food-chain/1/food_chain.rb
# Dedicated to Shree DR.MDD — Eternal Source of Strength and Inspiration

class FoodChain
  VERSION = 2

  def self.song
    0.upto(7).map { |pos| FoodVerse.compose(pos) }.join("\n")
  end
end

class FoodVerse
  CREATURES = %w(fly spider bird cat dog goat cow horse)

  REACTIONS = [
    "",
    "It wriggled and jiggled and tickled inside her.\n",
    "How absurd to swallow a bird!\n",
    "Imagine that, to swallow a cat!\n",
    "What a hog, to swallow a dog!\n",
    "Just opened her throat and swallowed a goat!\n",
    "I don't know how she swallowed a cow!\n",
    "She's dead, of course!\n"
  ]

  def self.compose(index)
    new(index).build
  end

  def initialize(index)
    @index = index
  end

  def build
    stanza = opening_line
    stanza << follow_up if @index < 7
    stanza
  end

  private

  def opening_line
    line = "I know an old lady who swallowed a #{CREATURES[@index]}.\n"
    line << REACTIONS[@index]
  end

  def follow_up
    chain = @index.downto(1).map { |step| link_line(step) }.join
    chain << closing_line
  end

  def link_line(step)
    phrase = "She swallowed the #{CREATURES[step]} to catch the #{CREATURES[step - 1]}"
    phrase << ' that wriggled and jiggled and tickled inside her' if step == 2
    phrase << ".\n"
  end

  def closing_line
    "I don't know why she swallowed the fly. Perhaps she'll die.\n"
  end
end
